kythe/rust/extractor/src/bin/extractor.rs

```diff
@@ -0,0 +1,230 @@
+        required_input,
+    );
+    let mut indexed_compilation_bytes: Vec<u8> = Vec::new();
+    indexed_compilation
+        .write_to_vec(&mut indexed_compilation_bytes)
+        .with_context(|| "Failed to serialize IndexedCompilation to bytes".to_string())?;
+    let indexed_compilation_digest = sha256digest(&indexed_compilation_bytes);
+    kzip_add_file(
+        format!("pbunits/{}", indexed_compilation_digest),
+        &indexed_compilation_bytes,
+        &mut kzip,
+    )?;
+
+    println!("Generated {:?}", &config.output_path);
+    Ok(())
+}
+
+// Inspired by https://gist.github.com/Igosuki/27e18369fec5a05d4019ac5d321e1779
+const SPAWN_INFO: protobuf::ext::ExtFieldOptional<
+    ExtraActionInfo,
+    protobuf::types::ProtobufTypeMessage<SpawnInfo>,
+> = protobuf::ext::ExtFieldOptional { field_number: 1003, phantom: ::std::marker::PhantomData };
+
+/// Attempt to parse ExtraActionInfo protobuf from file_path and extract the
+/// SpawnInfo extension
+fn get_spawn_info(file_path: impl AsRef<Path>) -> Result<SpawnInfo> {
+    let mut file = File::open(file_path).context("Failed to open extra action file")?;
+
+    let mut file_contents_bytes = Vec::new();
+    file.read_to_end(&mut file_contents_bytes).context("Failed to read extra action file")?;
+
+    let extra_action = protobuf::parse_from_bytes::<ExtraActionInfo>(&file_contents_bytes)
+        .context("Failed to parse extra action protobuf")?;
+
+    SPAWN_INFO.get(&extra_action).ok_or_else(|| anyhow!("SpawnInfo extension missing"))
+}
+
+/// Create an IndexedCompilation protobuf from the supplied arguments
+///
+/// * `source_files` - The names of the source files required by the build
+///   target
+/// * `arguments` - The arguments used to compile the target
+/// * `build_output_path` - The output path of the build target
+/// * `required_input` - The generated data for the CompilationUnit
+///   `required_input` field
+fn create_indexed_compilation(
+    source_files: Vec<String>,
+    arguments: Vec<String>,
+    build_output_path: String,
+    required_input: Vec<CompilationUnit_FileInput>,
+) -> IndexedCompilation {
+    let mut compilation_unit = CompilationUnit::new();
+
+    let mut vname = VName::new();
+    // TODO(Arm1stice): Determine proper Corpus/Root/Path for VName
+    vname.set_corpus("kythe".into());
+    vname.set_language("rust".into());
+    compilation_unit.set_v_name(vname);
+    compilation_unit.set_source_file(protobuf::RepeatedField::from_vec(source_files));
+    compilation_unit.set_argument(protobuf::RepeatedField::from_vec(arguments));
+    compilation_unit.set_required_input(protobuf::RepeatedField::from_vec(required_input));
+    compilation_unit.set_output_key(build_output_path);
+
+    let mut indexed_compilation = IndexedCompilation::new();
+    indexed_compilation.set_unit(compilation_unit);
+    indexed_compilation
+}
+
+/// Generate sha256 hex digest of a vector of bytes
+fn sha256digest(bytes: &[u8]) -> String {
```

Comment:
The result of this function should always be lowercase.

Reply:
The tests for the rust-crypto crate ensure that the `.result_str()` function always returns a lowercase hex string